File: grida-domains/domain-rds/src/main/java/org/grida/entity/user/UserEntity.java
package org.grida.entity.user;

import lombok.*;
import org.grida.domain.user.Gender;
import org.grida.domain.user.UserAppearance;
import org.grida.domain.user.UserRole;
import org.grida.entity.base.BaseEntity;

import javax.persistence.*;
import java.time.LocalDateTime;

@Entity
@Table(name = "users", indexes = {@Index(name = "idx_email", columnList = "email", unique = true)})
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor
public class UserEntity extends BaseEntity {

    @Column
    private String email;

    @Column(name = "encrypted_password")
    private String password;

    @Column
    @Enumerated(EnumType.STRING)
    private UserRole role;

    @Column
    private String nickname;

    @Column
    private Integer age;

    @Column
    @Enumerated(EnumType.STRING)
    private Gender gender;

    @Column
    private String hairStyle;

    @Column
    private String glasses;

    @Builder
    public UserEntity(
            LocalDateTime createdAt,
            LocalDateTime lastModifiedAt,
            String email, String password,
            UserRole role,
            String nickname
    ) {
        super(createdAt, lastModifiedAt);
        this.email = email;
        this.password = password;
        this.role = role;
        this.nickname = nickname;
    }

    public void modifyAppearance(UserAppearance appearance) {
        this.age = appearance.age();
        this.gender = appearance.gender();
        this.hairStyle = appearance.hairStyle();
        this.glasses = appearance.glasses();
    }
}

File: grida-database/database-rds/src/main/kotlin/org/grida/persistence/user/UserEntity.kt
package org.grida.persistence.user

import org.grida.domain.user.Role
import org.grida.domain.user.User
import org.grida.persistence.base.BaseEntity
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.EnumType
import javax.persistence.Enumerated
import javax.persistence.GeneratedValue
import javax.persistence.GenerationType
import javax.persistence.Id
import javax.persistence.Table

@Entity
@Table(name = "users")
class UserEntity(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "user_id")
    var id: Long = 0,

    var username: String,

    @Column(name = "encrypted_password")
    var password: String,

    @Enumerated(EnumType.STRING)
    var role: Role,

    var nickname: String
) : BaseEntity()
